src/lazy/text/matched.rs

```diff
@@ -154,3 +158,226 @@ impl MatchedFloat {
+        if !all_are_hex_digits {
+            return Err(IonError::Decoding(
+                DecodingError::new(format!(
+                    "found a {}-hex-digit escape sequence that contained an invalid hex digit",
+                    num_digits,
+                ))
+                .with_position(input.offset()),
+            ));
+        }
+        // We just confirmed all of the digits are ASCII hex digits, so these steps cannot fail.
+        let hex_digits = std::str::from_utf8(hex_digit_bytes).unwrap();
+        let code_point = u32::from_str_radix(hex_digits, 16).unwrap();
+
+        // Check to see if this is a high surrogate; if it is, our code point isn't complete. Another
+        // unicode escape representing the low surrogate has to be next in the input to complete it.
+        // See the docs for this helper function for details. (Note: this will only ever be true for
+        // 4- and 8-digit escape sequences. `\x` escapes don't have enough digits to represent a
+        // high surrogate.)
+        if code_point_is_a_high_surrogate(code_point) {
+            todo!("support surrogate pairs")
+        }
+
+        // A Rust `char` can represent any Unicode scalar value--a code point that is not part of a
+        // surrogate pair. If the value we found isn't a high surrogate, then it's a complete scalar
+        // value. We can safely convert it to a `char`.
+        let character = char::from_u32(code_point).unwrap();
+        let utf8_buffer: &mut [u8; 4] = &mut [0; 4];
+        let utf8_encoded = character.encode_utf8(utf8_buffer);
+        sanitized.extend_from_slice(utf8_encoded.as_bytes());
+
+        // Skip beyond the digits we just processed
+        Ok(input.slice_to_end(num_digits))
+    }
+}
+
+/// Returns `true` if the provided code point is a utf-16 high surrogate.
+///
+/// Terse primer: Unicode text is made up of a stream of unsigned integers called 'code points'.
+/// What a person might think of as a 'character' (for example: 'a', '本', or '🥸') can be made up
+/// of one or more code points.
+///
+/// A single code point can require up to 21 bits. Depending on which Unicode encoding you're using,
+/// these 21 bits can come with different amounts of additional overhead bits:
+/// * In utf-8, a code point can be 1, 2, 3, or 4 bytes, with some bits in each byte being used
+///   for the code point and others being used to indicate whether more bytes are coming.
+/// * In utf-16, a code point can be 2 bytes or 4 bytes. If it's four bytes, the first two bytes will
+///   be a 'high surrogate' (a value between 0xD800 and 0xDFFF) to communicate that another two
+///   bytes are coming to complete the code point.
+/// * In utf-32, a code point is always 32 bits. This is a bit wasteful, but makes for simple
+///   processing.
+///
+/// This helper function detects high surrogates (which are only used in utf-16) so the parser
+/// can know to require a second one immediately following.
```

Comment:
The specification mentions surrogate pair handling in a "MAY"-style statement.

> Ion does not specify the behavior of specifying invalid Unicode code points or surrogate code points (used only for UTF-16) using the escape sequences. It is highly recommended that Ion implementations reject such escape sequences as they are not proper Unicode as specified by the standard. To this point, consider the Ion string sequence, "\uD800\uDC00". A compliant parser may throw an exception because surrogate characters are specified outside of the context of UTF-16, accept the string as a technically invalid sequence of two Unicode code points (i.e. U+D800 and U+DC00), or interpret it as the single Unicode code point U+00010000. In this regard, the Ion string data type does not conform to the Unicode specification. A strict Unicode implementation of the Ion text should not accept such sequences.

So you're right that they're not required (and even actively discouraged). However,`ion-java` supports them and `ion-tests` has tests for them. The existing `ion-rust` text reader also supports them. I'm not in a rush to implement this and could be convinced to formalize it as an error case instead.